Add net8.0 and use Regex source generator

martincostello opened the pull request:
- Add target for `net8.0` and update to C# 12.
- Use the Regex source generator via `[GeneratedRegex]` when targeting `net8.0`.
- Bump the checkout and setup-dotnet actions to v4 to fix warnings about deprecated Node.js versions.
- Remove invalid key from NuGet publishing workflow.

martincostello committed:
Fix invalid workflow

`workflow_dispatch` cannot have a `branches` key.

martincostello committed:
Bump GitHub actions

Bump the checkout and setup-dotnet actions to v4.

martincostello committed:
Target net8.0

Add target for `net8.0` and update to C# 12.

martincostello committed:
Bump StyleCop.Analyzers

Bump StyleCop.Analyzers to support C# 12.

martincostello committed:
Use Regex source generator

Use the Regex source generator via `[GeneratedRegex]` when targeting `net8.0`.

goelhardik commented on Directory.Build.props:
```diff
@@ -37,6 +37,6 @@
   <ItemGroup>
     <PackageReference Include="Nerdbank.GitVersioning" Version="3.3.37" PrivateAssets="All" />
-    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118" PrivateAssets="All" Condition="'$(StyleCopEnabled)'=='true'" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.556" PrivateAssets="All" Condition="'$(StyleCopEnabled)'=='true'" />
```
I think we will need to use a release version here, not a beta version so that the final package doesn't have to be marked prerelease. Is there any particular reason to use a beta version here?

martincostello replied:
There hasn't been a non-stable version for years now, and only this version understands recent C# 8+ with sensible formatting. I don't know why they won't commit to a new stable release, but they haven't. I use this beta version in all my own and work protects.

martincostello replied:
Also, this package is only a build time dependency, it should not be a dependency of the shipped package.

goelhardik replied:
Yeah sounds good then.

martincostello committed:
Update NerdBank.GitVersioning

Bump to see if it fixes the Docker build.

goelhardik commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-FROM mcr.microsoft.com/dotnet/sdk:6.0 AS builder
```
I think we should test with both > 8.0 and < 8.0 versions to make sure both cases are covered.

martincostello replied:
The default behaviour of `dotnet test` will already run for all runtimes specified:

https://github.com/martincostello/ignore/blob/8bcb0c99cbf0448171d57e074a71d884eb4ed658/src/Ignore.Tests/Ignore.Tests.csproj#L4

When I run it locally on Windows, you can see it tests both `net6.0` and `net8.0`.

I just need to fix the Dockerfile so it works.

martincostello replied:
CI now tests [.NET 6](https://github.com/goelhardik/ignore/actions/runs/9988158458/job/27604014766#step:3:131) and [.NET 8](https://github.com/goelhardik/ignore/actions/runs/9988158458/job/27604014766#step:3:226).

goelhardik replied:
Awesome

martincostello committed:
Fix Docker warnings

- Use JSON format for arguments.
- Use key=value format for environment variables.

martincostello committed:
Install .NET 6 runtime

Install the .NET 6 runtime for use with the tests.

martincostello committed:
Fix tests

Add `--` to pass coverage flag through to tests, otherwise it thinks it's a runtime (`--runtime`).

martincostello committed:
Update test dependencies

Update tests to use up-to-date dependencies (except `System.IO.Abstractions`).

martincostello committed:
Remove exec

See if removing exec fixes the build.

martincostello committed:
Update README

- Update the TFMs noted in the README.
- Fix markdownlint warnings.

martincostello commented:
Is this worth changing to `0.2` for the new TFM?

https://github.com/goelhardik/ignore/blob/5d53183feebe4430bd84050d9cc1494be53b59d4/version.json#L3

goelhardik commented:
> Is this worth changing to `0.2` for the new TFM?
> 
> https://github.com/goelhardik/ignore/blob/5d53183feebe4430bd84050d9cc1494be53b59d4/version.json#L3

@martincostello yeah I think it's a good idea, we can bump to 0.2

martincostello committed:
Fix coverage path

`-r` became `--runtime` in .NET 8, so need to use `--results-directory`.

martincostello committed:
Bump version

Bump version to 0.2.x.

goelhardik reviewed (approved):
Thanks @martincostello for making this upgrade.
